### PR TITLE
ENH: is_homogeneous

### PR DESCRIPTION
**Pull request by TomAugspurger**
Split https://github.com/pandas-dev/pandas/pull/22325

@jorisvandenbossche suggested moving this off of the BlockManager.

Right now, I've made this public. Do we want that? If so I'll add to api.rst, release note, etc. Otherwise, I'll make it private.

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
I would maybe rather go with private for now. Or do think of good practical use cases outside of our internals?

**Comment by TomAugspurger** [comment]
My main use-case would be in when writing a custom scikit-learn estimator. "Can I safely convert this DataFrame to an ndarray without converting to an ndarray of objects". Even then, it's not quite perfect as there are things like converting ints to floats, which may be OK there.

Regardless, let's keep it private for now, and open it up if there's demand..

**Comment by jorisvandenbossche** [comment]
Although some dtypes will still convert to object dtype of course (but that's maybe not your direct concern). Isn't it rather something like "all numeric dtypes" that you need there? (as indeed mixed int and float might not be a problem)

**Comment by TomAugspurger** [comment]
Yes, most likely.

For reference though, this is directly useful within pandas for determining whether we can safely do a cross-section on a DataFrame containing extension dtypes. If they're all the same type we want to take a special path. A PR implementing that is incoming, but this felt separate enough to be standalone.

**Comment by jorisvandenbossche** [comment]
> For reference though, this is directly useful within pandas for determining whether we can safely do a cross-section on a DataFrame containing extension dtypes

Yes, I know, but that seems more internal usage (therefore my question). 

But as you said, can always make it public later if there is demand / use case.

**Comment by jreback** [comment]
his should be internal only

File: pandas/core/indexes/multi.py [hunk 1]
```diff
@@ -288,6 +288,26 @@ def _verify_integrity(self, labels=None, levels=None):
     def levels(self):
         return self._levels
 
+    @property
+    def _is_homogeneous(self):
+        """Whether the levels of a MultiIndex all have the same dtype.
```
Comment (jreback) [inline pandas/core/indexes/multi.py]: can u share docstrings at all?

Comment (TomAugspurger) [reply pandas/core/indexes/multi.py]: I think they're all different enough that sharing would be burdensome.

**Comment by TomAugspurger** [comment]
Does this look OK? I'd like to get back to SparseArray today, which will depend on this and #22785.

File: pandas/core/frame.py [hunk 2]
```diff
@@ -613,6 +613,34 @@ def shape(self):
...
+        bool
+
+        Examples
+        --------
+        >>> DataFrame({"A": [1, 2], "B": [3, 4]})._is_homogeneous
+        True
+        >>> DataFrame({"A": [1, 2], "B": [3.0, 4.0]})._is_homogeneous
+        False
+
+        Items with the type but different sizes are considered different
```
Comment (jorisvandenbossche) [inline pandas/core/frame.py]: the type -> the same type

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jreback** [comment]
maybe name _is_homogeneous_type to be consistent?

otherwise lgtm

**Comment by TomAugspurger** [comment]
Unlike `is_*_dtype`, I don't think there's any ambiguity here as to whether this is an array or dtype that's being checked, so I prefer the shorter name if that's OK.

**Comment by jreback** [comment]
no it IS different, this is like is_mixed_type (and not dype things)

pls change this

**Comment by TomAugspurger** [comment]
Ahh, I see what you think it should be consistent with. Will fix that in #22785.

**Comment by jorisvandenbossche** [comment]
If we want to add a suffix, I personally find 'dtype' more logical as 'type', since it is the dtype that is checked to be homogenous

**Pull request merged**